app/controllers/concerns/api/token_authenticatable.rb

```diff
@@ -0,0 +1,33 @@
+module API
+  module TokenAuthenticatable
+    extend ActiveSupport::Concern
+    include ActionController::HttpAuthentication::Token::ControllerMethods
+
+    included do
+      before_action :authenticate
+      before_action :set_paper_trail_whodunnit
+    end
+
+  private
+
+    def authenticate
+      authenticate_token || render_unauthorized
+    end
+
+    def authenticate_token
+      authenticate_with_http_token do |unhashed_token|
+        @current_api_token = APIToken.find_by_unhashed_token(unhashed_token).tap do |api_token|
+          api_token.update!(last_used_at: Time.zone.now)
+        end
+      end
+    end
+
+    def render_unauthorized
+      render json: { error: I18n.t(:unauthorized) }.to_json, status: :unauthorized
+    end
+
+    def current_lead_provider
+      @current_api_token&.owner
+    end
```

Comment:
Instead of `current_user` we use `current_leader_provider` as we actually only return `LeadProvider`